Mathlib/AlgebraicGeometry/Morphisms/Basic.lean

```diff
@@ -156,347 +233,248 @@ instance targetAffineLocally_respectsIso (P : AffineTargetMorphismProperty)
+  respectsIso := inferInstance
+  to_basicOpen _ _ H := IsLocalAtTarget.restrict H _
+  of_basicOpenCover {_ Y} _ _ _ hs := IsLocalAtTarget.of_iSup_eq_top _
+    (((isAffineOpen_top Y).basicOpen_union_eq_self_iff _).mpr hs)
 
 /-- A `P : AffineTargetMorphismProperty` is stable under base change if `P` holds for `Y ⟶ S`
 implies that `P` holds for `X ×ₛ Y ⟶ X` with `X` and `S` affine schemes. -/
-def StableUnderBaseChange (P : AffineTargetMorphismProperty) : Prop :=
+def AffineTargetMorphismProperty.StableUnderBaseChange (P : AffineTargetMorphismProperty) : Prop :=
   ∀ ⦃X Y S : Scheme⦄ [IsAffine S] [IsAffine X] (f : X ⟶ S) (g : Y ⟶ S),
     P g → P (pullback.fst f g)
 #align algebraic_geometry.affine_target_morphism_property.stable_under_base_change AlgebraicGeometry.AffineTargetMorphismProperty.StableUnderBaseChange
 
-theorem IsLocal.targetAffineLocally_pullback_fst_of_right_of_stableUnderBaseChange
-    {P : AffineTargetMorphismProperty} (hP : P.IsLocal) (hP' : P.StableUnderBaseChange)
-    {X Y S : Scheme} (f : X ⟶ S) (g : Y ⟶ S) [IsAffine S] (H : P g) :
-    targetAffineLocally P (pullback.fst f g) := by
-  have := hP.1
-  -- Porting note: rewrite `(hP.affine_openCover_TFAE ...).out 0 1` doesn't work
-  have h01 := (hP.affine_openCover_TFAE (pullback.fst f g)).out 0 1
-  rw [h01]
-  use X.affineCover, inferInstance
+end AffineTargetMorphismProperty
+
+section targetAffineLocally
+
+/-- For a `P : AffineTargetMorphismProperty`, `targetAffineLocally P` holds for
+`f : X ⟶ Y` whenever `P` holds for the restriction of `f` on every affine open subset of `Y`. -/
+def targetAffineLocally (P : AffineTargetMorphismProperty) : MorphismProperty Scheme :=
+  fun {X Y : Scheme} (f : X ⟶ Y) => ∀ U : Y.affineOpens, P (f ∣_ U)
+#align algebraic_geometry.target_affine_locally AlgebraicGeometry.targetAffineLocally
+
+theorem of_targetAffineLocally_of_isPullback
+    {P : AffineTargetMorphismProperty} [P.IsLocal]
+    {U V X Y : Scheme.{u}} [IsAffine U] {f : X ⟶ Y} {g : U ⟶ Y} [IsOpenImmersion g]
+    {iV : V ⟶ X} {f' : V ⟶ U} (h : IsPullback iV f' f g) (hf : targetAffineLocally P f) :
+    P f' := by
+  rw [← P.cancel_left_of_respectsIso h.isoPullback.inv, h.isoPullback_inv_snd]
+  exact (P.arrow_mk_iso_iff
+    (morphismRestrictOpensRange f _)).mp (hf ⟨_, isAffineOpen_opensRange g⟩)
+
+instance (P : AffineTargetMorphismProperty) [P.toProperty.RespectsIso] :
+    (targetAffineLocally P).RespectsIso := by
+  constructor
+  · introv H U
+    rw [morphismRestrict_comp, P.cancel_left_of_respectsIso]
+    exact H U
+  · introv H
+    rintro ⟨U, hU : IsAffineOpen U⟩; dsimp
+    haveI : IsAffine _ := hU.preimage_of_isIso e.hom
+    rw [morphismRestrict_comp, P.cancel_right_of_respectsIso]
+    exact H ⟨(Opens.map e.hom.val.base).obj U, hU.preimage_of_isIso e.hom⟩
+
+/-- `IsLocalAtTarget P Q` is a type class asserting that `P` is local at target, and over affine
+  schemes, it is equivalent to `Q : AffineTargetMorphismProperty`. -/
```

Comment:
This seems to be a bit off, since `P` is not required to be local at the target in the definition. Maybe this is clearer?
```suggestion
/-- `HasAffineProperty P Q` is a type class asserting that over affine schemes `P` is equivalent to the local `Q : AffineTargetMorphismProperty`. -/
```

Reply:
I changed the docstring.